scheduler/launchd.go

```diff
@@ -0,0 +1,149 @@
+package scheduler
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"path"
+
+	homedir "github.com/mitchellh/go-homedir"
+)
+
+type launchd struct{}
+
+// Enable creates the launchd script and then loads it using launchctl
+func (ld *launchd) Enable() (launchdWasLoaded bool, err error) {
+	launchdAgentFilename, err := ld.GetAgentFilename()
+	if err != nil {
+		return false, ErrModifyingLaunchd{origError: err}
+	}
+
+	launchdWasLoaded, err = ld.enable(&fileFunctionsPassthrough{}, launchdAgentFilename)
+	if err != nil {
+		return launchdWasLoaded, ErrModifyingLaunchd{origError: err}
+	}
+	return launchdWasLoaded, nil
+}
+
+// Disable unloads the launchd script, then deletes the script
+func (ld *launchd) Disable() (launchdWasUnloaded bool, err error) {
+	launchdAgentFilename, err := ld.GetAgentFilename()
+	if err != nil {
+		return false, ErrModifyingLaunchd{origError: err}
+	}
+
+	launchdWasUnloaded, err = ld.disable(&fileFunctionsPassthrough{}, launchdAgentFilename)
+	if err != nil {
+		return launchdWasUnloaded, ErrModifyingLaunchd{origError: err}
+	}
+	return launchdWasUnloaded, nil
+}
+
+func (ld *launchd) Name() string {
+	return "launchd"
+}
+
+func (ld *launchd) enable(fileFunctions fileFunctionsInterface, launchdAgentFilename string) (
+	launchdFileWasCreated bool, err error) {
+
+	if _, err := fileFunctions.OsStat(launchdAgentFilename); os.IsNotExist(err) {
```

Comment:
and if `err` is something else, like `permission denied` ?

Reply:
`os.Stat` can only return errors of `*PathError` - this line is simply checking if the file exists to see whether we need to write it.

If there's a permission error, I presume it'd be caught by `launchctl`?

Reply:
Modified this, factoring out a function `fileExists`.